core/src/main/scala/kafka/admin/AdminClient.scala

```diff
@@ -15,38 +15,54 @@ package kafka.admin
 import java.nio.ByteBuffer
 import java.util.{Collections, Properties}
 import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.{TimeUnit, Future}
 
+import kafka.admin.AdminClient.PurgeDataResult
 import org.apache.kafka.common.requests.ApiVersionsResponse.ApiVersion
 import kafka.common.KafkaException
 import kafka.coordinator.GroupOverview
 import kafka.utils.Logging
 import org.apache.kafka.clients._
-import org.apache.kafka.clients.consumer.internals.{ConsumerNetworkClient, ConsumerProtocol, RequestFuture}
+import org.apache.kafka.clients.consumer.internals.{RequestFutureAdapter, ConsumerNetworkClient, ConsumerProtocol, RequestFuture}
 import org.apache.kafka.common.config.ConfigDef.{Importance, Type}
 import org.apache.kafka.common.config.{AbstractConfig, ConfigDef}
 import org.apache.kafka.common.metrics.Metrics
 import org.apache.kafka.common.network.Selector
 import org.apache.kafka.common.protocol.{ApiKeys, Errors}
 import org.apache.kafka.common.requests._
 import org.apache.kafka.common.requests.OffsetFetchResponse
-import org.apache.kafka.common.utils.{Time, Utils}
+import org.apache.kafka.common.utils.{KafkaThread, Time, Utils}
 import org.apache.kafka.common.{Cluster, Node, TopicPartition}
 
 import scala.collection.JavaConverters._
 import scala.util.Try
 
 class AdminClient(val time: Time,
                   val requestTimeoutMs: Int,
+                  val retryBackoffMs: Long,
                   val client: ConsumerNetworkClient,
                   val bootstrapBrokers: List[Node]) extends Logging {
 
+  @volatile var running: Boolean = true
+
+  val networkThread = new KafkaThread("admin-client-network-thread", new Runnable {
+    override def run() {
+      while (running) {
+        client.poll(retryBackoffMs)
+        // Avoid busy loop so that user thread can get the lock on the ConsumerNetworkClient to enqueue requests.
+        if (client.pendingRequestCount() == 0)
+          Thread.sleep(retryBackoffMs)
```

Comment:
Since `ConsumerNetworkClient` is supposed to be thread-safe, I think it will be useful to improve `ConsumerNetworkClient` to provide better performance when multiple threads are calling `poll(..)` and `send(...)`. But I would prefer not to make big change for `ConsumerNetworkClient` in this patch unless there is strong reason to do so.

I see your point with using a queue in AdminClient to pass request objects from user thread to the sender thread and have sender thread be the only thread that calls `poll(..)` and `send(..)`. While this can work, the code seems too complex. Because we also need another queue to pass the Future objects from sender thread to user thread so that user thread can wait on the Future object. I would prefer not to over-engineer Scala AdminClient because it will be replaced with Java AdminClient in the future.

I have changed the code so that it uses `Thread.yield` to increase the chance a user thread can get the lock of `ConsumerNetworkClient` and send the requests. Alternatively the sender thread can sleep for 1 ms after each poll to increase the chance of user thread get the lock. Note that if user thread calls `send(...)` while the sender thread is blocked on selected, the sender thread will be waken up as of current implementation.